src/main/host/descriptor/file.c

```diff
@@ -286,6 +297,15 @@ int file_openat(File* file, File* dir, const char* pathname, int flags, mode_t m
         file->type = FILE_TYPE_REGULAR;
     }
 
+    int originalFlags = flags;
+
+    // move any emulated flags from 'flags' to 'emulatedFlags'
```

Comment:
I think this is part of my confusion. I thought that `emulatedFlags` were the flags set on the file descriptor from the managed process's point of view, and that `originalFlags` was the set that Shadow had set on the native file descriptor. IIUC though with this change, the set of flags from the managed process's point of view is actually `originalFlags | emulatedFlags`.

This could be addressed with some explanation in the struct definition and/or header, but I think the way I *thought* it worked would actually be clearer. Is there a reason not do it that way?

Reply:
Shadow doesn't track the file flags for the `File*` objects. It only stores the original flags that were used when the file was first opened. If the plugin uses `fcntl` to change the file flags, Shadow calls `fcntl` directly on the native file descriptor and the stored flags are no longer up to date. I renamed `flags` to `originalFlags` to try to make it clearer that these flags are historical and shouldn't be used as the file's actual flags.

The idea of `emulatedFlags` is separate from this. Rather than passing every flag from the plugin to the native file descriptor as Shadow does now, there are some flags that shouldn't be passed through (for example `O_CLOEXEC`). The `emulatedFlags` field was meant to keep track of these flags that don't get passed through from the plugin to the native file descriptor.

I agree that I should improve the documentation and naming here.

> IIUC though with this change, the set of flags from the managed process's point of view is actually originalFlags | emulatedFlags.

The flags from the managed process' point of view is `(fcntl(nativeFd, FD_GETFL) & EMULATED_FLAG_MASK) | emulatedFlags`. Shadow doesn't know anything about the actual flags.